runtime/call_metrics.go

```diff
@@ -0,0 +1,276 @@
+	http.StatusUpgradeRequired:               true, // 426
+	http.StatusPreconditionRequired:          true, // 428
+	http.StatusTooManyRequests:               true, // 429
+	http.StatusRequestHeaderFieldsTooLarge:   true, // 431
+	http.StatusUnavailableForLegalReasons:    true, // 451
+	http.StatusInternalServerError:           true, // 500
+	http.StatusNotImplemented:                true, // 501
+	http.StatusBadGateway:                    true, // 502
+	http.StatusServiceUnavailable:            true, // 503
+	http.StatusGatewayTimeout:                true, // 504
+	http.StatusHTTPVersionNotSupported:       true, // 505
+	http.StatusVariantAlsoNegotiates:         true, // 506
+	http.StatusInsufficientStorage:           true, // 507
+	http.StatusLoopDetected:                  true, // 508
+	http.StatusNotExtended:                   true, // 510
+	http.StatusNetworkAuthenticationRequired: true, // 511
+}
+
+type inboundMetrics struct {
+	Recvd tally.Counter // inbound.calls.recvd
+}
+
+type outboundMetrics struct {
+	Sent tally.Counter // outbound.calls.sent
+}
+
+type commonMetrics struct {
+	logger  *zap.Logger
+	Latency tally.Timer   // [inbound|outbound].calls.latency
+	Success tally.Counter // [inbound|outbound].calls.success
+}
+
+type tchannelMetrics struct {
+	commonMetrics
+	AppErrors    tally.Counter // [inbound|outbound].calls.app-errors
+	SystemErrors tally.Counter // [inbound|outbound].calls.system-errors
+}
+
+type httpMetrics struct {
+	commonMetrics
+	Errors tally.Counter         // [inbound|outbound].calls.errors
+	Status map[int]tally.Counter // [inbound|outbound].calls.status.XXX
+}
+
+// InboundHTTPMetrics ...
+type InboundHTTPMetrics struct {
+	inboundMetrics
+	httpMetrics
+}
+
+// InboundTChannelMetrics ...
+type InboundTChannelMetrics struct {
+	inboundMetrics
+	tchannelMetrics
+}
+
+// OutboundHTTPMetrics ...
+type OutboundHTTPMetrics struct {
+	outboundMetrics
+	httpMetrics
+}
+
+// OutboundTChannelMetrics ...
+type OutboundTChannelMetrics struct {
+	outboundMetrics
+	tchannelMetrics
+}
+
+// NewInboundHTTPMetrics returns inbound HTTP metrics
+func NewInboundHTTPMetrics(scope tally.Scope, logger *zap.Logger) *InboundHTTPMetrics {
+	metrics := InboundHTTPMetrics{}
+	metrics.logger = logger
+	metrics.Recvd = scope.Counter(inboundCallsRecvd)
+	metrics.Latency = scope.Timer(inboundCallsLatency)
+	metrics.Success = scope.Counter(inboundCallsSuccess)
+	metrics.Errors = scope.Counter(inboundCallsErrors)
+	metrics.Status = make(map[int]tally.Counter, len(knownStatusCodes))
+	for statusCode := range knownStatusCodes {
+		metrics.Status[statusCode] = scope.Counter(fmt.Sprintf("%s.%d", inboundCallsStatus, statusCode))
+	}
+	return &metrics
+}
+
+// CollectMetrics for inbound HTTP calls
+func (m *InboundHTTPMetrics) CollectMetrics(startTime, finishTime time.Time, statusCode int) {
+	m.Recvd.Inc(1)
+	_, known := knownStatusCodes[statusCode]
+	if !known {
+		m.logger.Error(unknownStatusCodeMsg, zap.Int(unknownStatusCode, statusCode))
```

Comment:
please put the error string inline here so it's easier to find.

Reply:
please put the `unknownStatusCode` inline here as well so it's easier to read.

Reply:
Done.